### PR TITLE
respect contexts in a more timely manner

### PR DESCRIPTION
**Pull request by whyrusleeping**
This PR fixes the '7MB bursts' that @anarcat reported. While this fix works, if the lower calls 'hang', those goroutines *may* build up, but the contexts are cancelled so they should exit sooner rather than later. Shouldnt be an issue.

### CODE REVIEW

**Comment by anarcat** [comment]
i believe the 7.5MB bursts are documented in #872. i'll try to test this soon, if i can figure out how to test branches...

**Comment by whyrusleeping** [comment]
The main issue that this PR 'fixes' is that the network stack does not respect contexts. I think that if we switch over to using grpc, a lot of that will be fixed.

File: exchange/bitswap/bitswap.go [hunk 1]
```diff
@@ -385,7 +404,15 @@ func (bs *Bitswap) wantNewBlocks(ctx context.Context, bkeys []u.Key) {
...
-	wg.Wait()
+	done := make(chan struct{})
+	go func() {
+		wg.Wait()
+		close(done)
+	}()
+	select {
+	case <-done:
+	case <-ctx.Done():
+	}
```
Comment (jbenet) [inline exchange/bitswap/bitswap.go]: if bs.send hangs, we exit leaking a hung goroutine. 

the downside of this approach is that it  becomes easy to hide hanging things, and thus harder to find problems-- nodes would just slowly deteriorate. 

why is `bs.send` hanging? is there a way to speed that up and respect the context there, instead of bypassing it here?

---

For others following this along-- the codebase is full of examples of both things: places where we return asap ad let the slow-to-finish goroutine finish on its own, and places where we explicitly wait for it to be done, so that we know we've fully finished. there's tradeoffs and cases where we _must_ do one or the other.

File: exchange/bitswap/bitswap.go [hunk 2]
```diff
@@ -227,21 +227,40 @@ func (bs *Bitswap) HasBlock(ctx context.Context, blk *blocks.Block) error {
...
 	set := pset.New()
 	wg := sync.WaitGroup{}
-	for peerToQuery := range peers {
 
-		if !set.TryAdd(peerToQuery) { //Do once per peer
-			continue
-		}
+loop:
+	for {
+		select {
```
Comment (jbenet) [inline exchange/bitswap/bitswap.go]: i think this select is definitely unavoidable-- we want to bail out even if `peers` has more.

File: exchange/bitswap/bitswap.go (hunk 1 above)
Comment (whyrusleeping) [inline exchange/bitswap/bitswap.go]: send is hanging because of the second stack trace in this gist: https://gist.github.com/whyrusleeping/a04c7bcf791d13e6ce73

Comment (whyrusleeping) [inline exchange/bitswap/bitswap.go]: basically comes down to our network stack not respecting contexts that well (but gRPC does!!!)

File: exchange/bitswap/bitswap.go [hunk 3]
```diff
@@ -227,21 +227,40 @@ func (bs *Bitswap) HasBlock(ctx context.Context, blk *blocks.Block) error {
...
+			if !ok {
+				break loop
+			}
 
-		wg.Add(1)
-		go func(p peer.ID) {
-			defer wg.Done()
-			if err := bs.send(ctx, p, m); err != nil {
-				log.Debug(err) // TODO remove if too verbose
+			if !set.TryAdd(peerToQuery) { //Do once per peer
```
Comment (jbenet) [inline exchange/bitswap/bitswap.go]: may be worth renaming `set.TryAdd` to `set.AddOnce(peer)` or `set.AddNew(peer)` or something. `set.TryAdd(peer)` isn't the clearest thing ever.

maybe another way to do it is:

```
// say set.Add(peer) returns an int of the number of times peer has been added

if set.Add(peer) == 1 {
  // do once per peer.
}
```

File: exchange/bitswap/bitswap.go [hunk 4]
```diff
@@ -227,21 +227,40 @@ func (bs *Bitswap) HasBlock(ctx context.Context, blk *blocks.Block) error {
...
+	}
+	done := make(chan struct{})
+	go func() {
+		wg.Wait()
+		close(done)
+	}()
+
+	select {
+	case <-done:
+	case <-ctx.Done():
```
Comment (jbenet) [inline exchange/bitswap/bitswap.go]: Same problem as in https://github.com/jbenet/go-ipfs/pull/900/files#r26015119

File: exchange/bitswap/bitswap.go (hunk 1 above)
Comment (jbenet) [inline exchange/bitswap/bitswap.go]: So-- when we move to gRPC we could revert this change and go back to waiting?

Also, any idea how long those goroutines wait in yamux before exiting? or what's making them wait in the first place? (worried about letting them build up indefinitely, could cause more problems)

Comment (whyrusleeping) [inline exchange/bitswap/bitswap.go]: I have no idea at all what yamux is doing. the method is `waitForSendErr` which selects on the error channel and the shutdown channel and exits when either even happens...

File: exchange/bitswap/bitswap.go (hunk 2 above)
Comment (whyrusleeping) [inline exchange/bitswap/bitswap.go]: yeah, agreed.

File: exchange/bitswap/bitswap.go (hunk 3 above)
Comment (whyrusleeping) [inline exchange/bitswap/bitswap.go]: Yeah... its a little confusing... i agree

File: exchange/bitswap/bitswap.go (hunk 1 above)
Comment (jbenet) [inline exchange/bitswap/bitswap.go]: hmm wonder if it's waiting for some sort of ack from the remote side...

File: exchange/bitswap/bitswap.go (hunk 4 above)
Comment (whyrusleeping) [inline exchange/bitswap/bitswap.go]: Yeah, but I think its okay because these goroutines dont hang forever (in my experience), just for a couple seconds while they wait for something to time out. So, I think its acceptable to throw them off into the void with the expectation that they finish moderately soon. Otherwise, we greatly limit ourselves in terms of bandwidth

File: exchange/bitswap/bitswap.go (hunk 1 above)
Comment (jbenet) [inline exchange/bitswap/bitswap.go]: - https://github.com/jbenet/go-peerstream/blob/master/transport/yamux/yamux.go
- https://github.com/hashicorp/yamux
- https://godoc.org/github.com/hashicorp/yamux

cant readily see anything im doing wrong. maybe it does wait for an ack to do flowcontrol / backpressure.

File: exchange/bitswap/bitswap.go (hunk 4 above)
Comment (jbenet) [inline exchange/bitswap/bitswap.go]: how would we detect such a build up? (other then OOMing the clients, which we should really stop doing).

Comment (whyrusleeping) [inline exchange/bitswap/bitswap.go]: http://xkcd.com/1495/

Comment (whyrusleeping) [inline exchange/bitswap/bitswap.go]: (clients adapt)
